hadoop-ozone/ozone-manager/src/test/java/org/apache/hadoop/ozone/om/service/TestSnapshotDeletingService.java

```diff
@@ -0,0 +1,233 @@
+    conf.setTimeDuration(OZONE_SNAPSHOT_DELETION_SERVICE_INTERVAL,
+        1000, TimeUnit.MILLISECONDS);
+    conf.setTimeDuration(OZONE_SNAPSHOT_DELETION_SERVICE_TIMEOUT,
+        100000, TimeUnit.MILLISECONDS);
+    conf.setQuietMode(false);
+    omTestManagers = new OmTestManagers(conf);
+    keyManager = omTestManagers.getKeyManager();
+    omMetadataManager = omTestManagers.getMetadataManager();
+    writeClient = omTestManagers.getWriteClient();
+    om = omTestManagers.getOzoneManager();
+  }
+
+  @After
+  public void cleanup() throws Exception {
+    om.stop();
+  }
+
+  @Test
+  public void testSnapshotKeySpaceReclaim() throws Exception {
+    int snapshotCount = 0;
+
+    SnapshotDeletingService snapshotDeletingService = (SnapshotDeletingService)
+        keyManager.getSnapshotDeletingService();
+
+    snapshotDeletingService.suspend();
+
+    OmKeyArgs key1 = createVolumeBucketKey(VOLUME_NAME, BUCKET_NAME,
+        BucketLayout.DEFAULT, "key1");
+
+    createSnapshot(VOLUME_NAME, BUCKET_NAME, "snap1", ++snapshotCount);
+
+    OmKeyArgs key2 = createKey(VOLUME_NAME, BUCKET_NAME, "key2");
+
+    // Key 1 cannot be deleted as it is still referenced by Snapshot 1.
+    writeClient.deleteKey(key1);
+    //Key 2 is deleted here, which means we can reclaim
+    // it when snapshot 2 is deleted.
+    writeClient.deleteKey(key2);
+
+    createSnapshot(VOLUME_NAME, BUCKET_NAME, "snap2", ++snapshotCount);
+    createKey(VOLUME_NAME, BUCKET_NAME, "key4");
+    OmKeyArgs key5 = createKey(VOLUME_NAME, BUCKET_NAME, "key5");
+    writeClient.deleteKey(key5);
+
+    createSnapshot(VOLUME_NAME, BUCKET_NAME, "snap3", ++snapshotCount);
+
+
+    String snapshotKey2 = "/vol1/bucket1/snap2";
+    SnapshotInfo snapshotInfo = om.getMetadataManager()
+        .getSnapshotInfoTable().get(snapshotKey2);
+
+    snapshotInfo
+        .setSnapshotStatus(SnapshotInfo.SnapshotStatus.SNAPSHOT_DELETED);
+    om.getMetadataManager()
+        .getSnapshotInfoTable().put(snapshotKey2, snapshotInfo);
+    snapshotInfo = om.getMetadataManager()
+        .getSnapshotInfoTable().get(snapshotKey2);
+    assertEquals(snapshotInfo.getSnapshotStatus(),
+        SnapshotInfo.SnapshotStatus.SNAPSHOT_DELETED);
+
+    snapshotDeletingService.resume();
+
+    GenericTestUtils.waitFor(() ->
+            snapshotDeletingService.getSuccessfulRunCount() >= 1,
+        1000, 10000);
+
+    OmSnapshot nextSnapshot = (OmSnapshot) om.getOmSnapshotManager()
+        .checkForSnapshot(VOLUME_NAME, BUCKET_NAME, getSnapshotPrefix("snap3"));
+
+    //Check key1 added to next non deleted snapshot db.
+    RepeatedOmKeyInfo omKeyInfo =
+        nextSnapshot.getMetadataManager()
+            .getDeletedTable().get("/vol1/bucket1/key1");
+    assertNotNull(omKeyInfo);
+
+    //Check key2 added active db as it can be reclaimed.
+    RepeatedOmKeyInfo omKeyInfo1 = omMetadataManager
+        .getDeletedTable().get("/vol1/bucket1/key2");
+
+    assertNotNull(omKeyInfo1);
+
+  }
```

Comment:
New UT request: Take snapshot on two buckets separately

Reply:
Relevant UT for reference: https://github.com/apache/ozone/pull/4273/files#diff-71c1f01e984fc77a5d653be56ec27d6347b2f8104c717a936b6a53a25c5116e3R602-R627